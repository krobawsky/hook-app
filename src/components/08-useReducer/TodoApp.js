import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';

import './style.css';
import { TodoList } from './TodoList';
import { TodoForm } from './TodoForm';

const init = () => {

    return JSON.parse(localStorage.getItem('todos')) || [];
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'learn react',
    //     done: false
    // }];

}

export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleAdd = (newTodo) => {
        
        dispatch({
            type: 'add',
            payload: newTodo
        });

    }

    const handleDelete = (todoId) => {
        
        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch(action);

    }

    const handleToggle = (todoId) => {

        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    return (
        <div>
            <h1>TodoApp ({ todos.length })</h1>
            <hr/>

            <div className="row">
                <div className="col-7">
                    <TodoList 
                        todos={ todos } 
                        handleDelete={ handleDelete } 
                        handleToggle={ handleToggle }
                    />
                </div>
                <div className="col-5">
                    <TodoForm 
                        handleAdd={ handleAdd }
                    />
                </div>
            </div>

        </div>
    )
}
