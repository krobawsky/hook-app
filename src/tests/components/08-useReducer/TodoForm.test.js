import React from 'react';
import { shallow } from 'enzyme';
import { TodoForm } from '../../../components/08-useReducer/TodoForm';

describe('Pruebas en <TodoForm />', () => {
    
    const handleAdd = jest.fn();

    const wrapper = shallow(
        <TodoForm
            handleAdd={ handleAdd }
        />
    );

    test('debe mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot();

    });

    test('NO debe de llamar el handleAdd', () => {
        
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} }); // simunlar submit

        expect( handleAdd ).toHaveBeenCalledTimes(0);

    });

    test('debe de llamar el handleAdd', () => {
        
        const value = 'Aprender xxx';
        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        });

        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({ preventDefault(){} }); // simunlar submit

        expect( handleAdd ).toHaveBeenCalledTimes(1);
        expect( handleAdd ).toHaveBeenCalledWith( expect.any(Object) );
        expect( handleAdd ).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false
        });

        expect( wrapper.find('input').prop('value') ).toBe('');

    });
    
})