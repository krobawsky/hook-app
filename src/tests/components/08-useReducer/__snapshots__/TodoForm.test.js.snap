// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoForm /> debe mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      name="description"
      onChange={[Function]}
      placeholder="Aprender ..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-2 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
